pkg/storage/storage.go

```diff
@@ -80,69 +97,157 @@ func (p *storage) GetContainerMetrics(pods ...apitypes.NamespacedName) ([]api.Ti
 
 	timestamps := make([]api.TimeInfo, len(pods))
 	resMetrics := make([][]metrics.ContainerMetrics, len(pods))
+	for podIdx, pod := range pods {
+		contPoints, present := p.pods[pod]
+		if !present {
+			continue
+		}
 
-	for i, pod := range pods {
-		metricPoint, present := p.pods[pod]
+		_, present = p.prevPods[pod]
 		if !present {
 			continue
 		}
 
-		contMetrics := make([]metrics.ContainerMetrics, len(metricPoint.Containers))
-		var earliestTS *time.Time
-		for i, contPoint := range metricPoint.Containers {
-			contMetrics[i] = metrics.ContainerMetrics{
+		var (
+			contMetrics = make([]metrics.ContainerMetrics, len(contPoints))
+			earliestTS  time.Time
+			window      time.Duration
+			contIdx     int
+		)
+		for contName, contPoint := range contPoints {
+			prevContPoint, present := p.prevPods[pod][contName]
+			if !present {
+				continue
+			}
+
+			cpuUsage := cpuUsageOverTime(contPoint.MetricsPoint, prevContPoint.MetricsPoint)
+			contMetrics[contIdx] = metrics.ContainerMetrics{
 				Name: contPoint.Name,
 				Usage: corev1.ResourceList{
-					corev1.ResourceName(corev1.ResourceCPU):    contPoint.CpuUsage,
+					corev1.ResourceName(corev1.ResourceCPU):    cpuUsage,
 					corev1.ResourceName(corev1.ResourceMemory): contPoint.MemoryUsage,
 				},
 			}
-			if earliestTS == nil || earliestTS.After(contPoint.Timestamp) {
-				ts := contPoint.Timestamp // copy to avoid loop iteration variable issues
-				earliestTS = &ts
+			if earliestTS.IsZero() || earliestTS.After(contPoint.Timestamp) {
+				earliestTS = contPoint.Timestamp
+				window = contPoint.Timestamp.Sub(prevContPoint.Timestamp)
 			}
+			contIdx++
 		}
-		if earliestTS == nil {
-			// we had no containers
-			earliestTS = &time.Time{}
-		}
-		timestamps[i] = api.TimeInfo{
-			Timestamp: *earliestTS,
-			Window:    kubernetesCadvisorWindow,
+		resMetrics[podIdx] = contMetrics
+		timestamps[podIdx] = api.TimeInfo{
+			Timestamp: earliestTS,
+			Window:    window,
 		}
-		resMetrics[i] = contMetrics
 	}
 	return timestamps, resMetrics, nil
 }
 
 func (p *storage) Store(batch *MetricsBatch) {
-	newNodes := make(map[string]NodeMetricsPoint, len(batch.Nodes))
-	var nodeCount, containerCount int
-	for _, nodePoint := range batch.Nodes {
+	p.storeNodeMetrics(batch.Nodes)
+	p.storePodMetrics(batch.Pods)
+}
+
+func (p *storage) storeNodeMetrics(nodes []NodeMetricsPoint) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	newNodes := make(map[string]NodeMetricsPoint, len(nodes))
+	prevNodes := make(map[string]NodeMetricsPoint, len(nodes))
+	var nodeCount int
+	for _, nodePoint := range nodes {
 		if _, exists := newNodes[nodePoint.Name]; exists {
 			klog.Errorf("duplicate node %s received", nodePoint.Name)
 			continue
 		}
-		nodeCount += 1
 		newNodes[nodePoint.Name] = nodePoint
+		nodeCount++
```

Comment:
Would be great to count only nodes that we can return metrics for. With those changes we should report node if we have both current and previous point.

Reply:
I replaced how we used to count by just adding `len(prev)` to the count so that it will only account for the resources for which we have 2 metric points.